File: Parallel-Csharp/src/Parallel/Linq101/MyLinq/MaxMethods.cs
using Parallel.Linq101.DataSources;

namespace Parallel.Linq101.MyLinq;

public class MaxMethod {
    //original:4 target:40
    public List<Product> GetProductList() => Products.ProductList;
    public List<Customer> GetCustomerList() => Customers.CustomerList;

    //max syntax : 5

    public int NumbersMax() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int maxNum = numbers.Max();
        return maxNum;
    }

    public int BigNumbersMax() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int maxNum = numbers.Where(n => n > 5).Max();
        return maxNum;
    }

    public int OddNumbersMax() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int maxNum = numbers.Where(n => n % 2 == 1).Max();
        return maxNum;
    }

    public double DoubleMax() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double maxNum = numbers.Max();
        return maxNum;
    }

    public double BigDoubleNumbersMax() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double maxNum = numbers.Where(d => d > 5).Max();
        return maxNum;
    }


    //max projection: 15

    public int NumbersMaxTwice() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int maxNum = numbers.Max(n => n * 2);
        return maxNum;
    }

    public int BigNumbersMaxTwice() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int maxNum = numbers.Where(n => n > 5).Max(n => n * 2);
        return maxNum;
    }

    public int OddNumbersMaxTwice() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int maxNum = numbers.Where(n => n % 2 == 1).Max(n => n * 2);
        return maxNum;
    }

    public int StringMaxLength() {
        string[] words = { "cherry", "apple", "blueberry" };
        int maxLength = words.Max(w => w.Length);
        return maxLength;
    }

    public int ShortStringMaxLength() {
        string[] words = { "cherry", "apple", "blueberry" };
        int maxLength = words.Where(w => w.Length < 7).Max(w => w.Length);
        return maxLength;
    }

    public double ProductUnitPriceMax() {
        List<Product> products = GetProductList();
        double maxUnitPrice = products.Max(p => p.UnitPrice);
        return maxUnitPrice;
    }

    public double ExpensiveProductUnitPriceMax() {
        List<Product> products = GetProductList();
        double maxUnitPrice = products.Where(p => p.UnitPrice > 100).Max(p => p.UnitPrice);
        return maxUnitPrice;
    }

    public int ExpensiveProductStockMax() {
        List<Product> products = GetProductList();
        int maxUnitPrice = products.Where(p => p.UnitPrice > 100).Max(p => p.UnitsInStock);
        return maxUnitPrice;
    }

    public int ProductLongNameLengthMax() {
        List<Product> products = GetProductList();
        int maxUnitPrice = products.Select(p => p.ProductName)
            .Where(name => name.Length > 7).Max(name => name.Length);
        return maxUnitPrice;
    }

    public List<Tuple<string, double>> CustomerOrdersTotalMax() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersTotalMax = customers.Select(c => Tuple.Create(c.CustomerId,
            c.Orders.Max(o => o.Total))).ToList();
        return ordersTotalMax;
    }

    public Dictionary<string, double> CustomerOrdersTotalMaxDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersTotalMax = customers.ToDictionary(c => c.CustomerId,
            c => c.Orders.Max(o => o.Total));
        return ordersTotalMax;
    }

    public List<Tuple<string, double>> CustomerInRegionOrdersTotalMax(string region) {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersTotalMax = customers.Where(c => c.Region.Equals(region))
            .Select(c => Tuple.Create(c.CustomerId,
                c.Orders.Max(o => o.Total))).ToList();
        return ordersTotalMax;
    }

    public Dictionary<string, double> CustomerInRegionOrdersTotalMaxDict(string region) {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersTotalMax = customers.Where(c => c.Region.Equals(region))
            .ToDictionary(c => c.CustomerId,
                c => c.Orders.Max(o => o.Total));
        return ordersTotalMax;
    }

    public List<Tuple<string, double>> CustomerExpensiveOrdersTotalMax() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersTotalMax = customers.Select(c => Tuple.Create(c.CustomerId,
            c.Orders.Where(o => o.Total > 100).Max(o => o.Total))).ToList();
        return ordersTotalMax;
    }

    public Dictionary<string, double> CustomerExpensiveOrdersTotalMaxDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersTotalMax = customers.ToDictionary(c => c.CustomerId,
            c => c.Orders.Where(o => o.Total > 100).Max(o => o.Total));
        return ordersTotalMax;
    }

    //max grouped: 10

    public List<Tuple<string, int>> ProductEachCategoryStock() {
        List<Product> products = GetProductList();
        List<Tuple<string, int>> categories = products.GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key, g.Max(p => p.UnitsInStock))).ToList();
        return categories;
    }

    public Dictionary<string, int> ProductEachCategoryStockDict() {
        List<Product> products = GetProductList();
        Dictionary<string, int> categories = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g => g.Max(p => p.UnitsInStock));
        return categories;
    }

    public List<Tuple<string, int>> ExpensiveProductEachCategoryStock() {
        List<Product> products = GetProductList();
        List<Tuple<string, int>> categories = products.GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key, g.Where(p => p.UnitPrice > 100)
                .Max(p => p.UnitsInStock))).ToList();
        return categories;
    }

    public Dictionary<string, int> ExpensiveProductEachCategoryStockDict() {
        List<Product> products = GetProductList();
        Dictionary<string, int> categories = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g =>
                g.Where(p => p.UnitPrice > 100).Max(p => p.UnitsInStock));
        return categories;
    }

    public List<Tuple<string, int>> CustomerEachRegionOrdersMax() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, int>> ordersMax = customers.GroupBy(c => c.Region)
            .Select(g => Tuple.Create(g.Key,
                g.Max(c => c.Orders.Length))).ToList();
        return ordersMax;
    }

    public Dictionary<string, int> CustomerEachRegionOrdersMaxDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, int> ordersMax = customers.GroupBy(c => c.Region)
            .ToDictionary(g => g.Key,
                g => g.Max(c => c.Orders.Length));
        return ordersMax;
    }

    public List<Tuple<string, int>> CustomerEachCountryOrdersMax() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, int>> ordersMax = customers.GroupBy(c => c.Country)
            .Select(g => Tuple.Create(g.Key,
                g.Max(c => c.Orders.Length))).ToList();
        return ordersMax;
    }

    public Dictionary<string, int> CustomerEachCountryOrdersMaxDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, int> ordersMax = customers.GroupBy(c => c.Country)
            .ToDictionary(g => g.Key,
                g => g.Max(c => c.Orders.Length));
        return ordersMax;
    }

    public List<Tuple<string, double>> CustomerEachRegionOrdersTotalMax() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersTotalMax = customers.GroupBy(c => c.Region)
            .Select(g => Tuple.Create(g.Key,
                g.SelectMany(c => c.Orders).Max(o => o.Total))).ToList();
        return ordersTotalMax;
    }

    public Dictionary<string, double> CustomerEachRegionOrdersTotalMaxDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersTotalMax = customers.GroupBy(c => c.Region)
            .ToDictionary(g => g.Key,
                g => g.SelectMany(c => c.Orders).Max(o => o.Total));
        return ordersTotalMax;
    }
    
    //max each group: 10
    public List<Tuple<string, List<Product>>> ProductEachCategoryMaxUnitPriceProducts() {
        List<Product> products = GetProductList();
        List<Tuple<string, List<Product>>> categories = products.GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key,
                g.Where(p => p.UnitPrice == g.Max(p => p.UnitPrice))
                    .ToList())).ToList();
        return categories;
    }
    
    public Dictionary<string, List<Product>> ProductEachCategoryMaxUnitPriceProductsDict() {
        List<Product> products = GetProductList();
        Dictionary<string, List<Product>> categories = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key,
                g=> g.Where(p => p.UnitPrice == g.Max(p => p.UnitPrice))
                    .ToList());
        return categories;
    }
    
    public List<Tuple<string, List<Product>>> ExpensiveProductEachCategoryMaxUnitPriceProducts() {
        List<Product> products = GetProductList();
        List<Tuple<string, List<Product>>> categories = products.Where(p=>p.UnitPrice>100).GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key,
                g.Where(p => p.UnitPrice == g.Max(p => p.UnitPrice))
                    .ToList())).ToList();
        return categories;
    }
    
    public Dictionary<string, List<Product>> ExpensiveProductEachCategoryMaxUnitPriceProductsDict() {
        List<Product> products = GetProductList();
        Dictionary<string, List<Product>> categories = products.Where(p=>p.UnitPrice>100).GroupBy(p => p.Category)
            .ToDictionary(g => g.Key,
                g=> g.Where(p => p.UnitPrice == g.Max(p => p.UnitPrice))
                    .ToList());
        return categories;
    }
    
    public List<Tuple<string, List<Customer>>> CustomerEachRegionOrdersMaxCustomer() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, List<Customer>>> customerOrdersMax = customers.GroupBy(c => c.Region)
            .Select(g => Tuple.Create(g.Key,
                g.Where(c => c.Orders.Length ==g.Max(c => c.Orders.Length)).ToList())).ToList();
        return customerOrdersMax;
    }

    public Dictionary<string, List<Customer>> CustomerEachRegionOrdersMaxCustomerDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, List<Customer>> customerOrdersMax = customers.GroupBy(c => c.Region)
            .ToDictionary(g => g.Key,
                g => g.Where(c => c.Orders.Length ==g.Max(c => c.Orders.Length)).ToList());
        return customerOrdersMax;
    }

    public List<Tuple<string, List<Customer>>> CustomerEachCountryOrdersMaxCustomer() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, List<Customer>>> customerOrdersMax = customers.GroupBy(c => c.Country)
            .Select(g => Tuple.Create(g.Key,
                g.Where(c => c.Orders.Length ==g.Max(c => c.Orders.Length)).ToList())).ToList();
        return customerOrdersMax;
    }

    public Dictionary<string, List<Customer>> CustomerEachCountryOrdersMaxCustomerDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, List<Customer>> customerOrdersMax = customers.GroupBy(c => c.Country)
            .ToDictionary(g => g.Key,
                g => g.Where(c => c.Orders.Length ==g.Max(c => c.Orders.Length)).ToList());
        return customerOrdersMax;
    }

    public List<Tuple<string, List<Order>>> CustomerEachRegionOrdersTotalMaxCustomer() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, List<Order>>> customerOrdersTotalMax = customers.GroupBy(c => c.Region)
            .Select(g => Tuple.Create(g.Key,
                g.SelectMany(c=>c.Orders).Where(
                    o=>o.Total == g.SelectMany(c=>c.Orders).Max(o => o.Total)).ToList())).ToList();
        return customerOrdersTotalMax;
    }

    public Dictionary<string, List<Order>> CustomerEachRegionOrdersTotalMaxCustomerDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, List<Order>> customerOrdersTotalMax = customers.GroupBy(c => c.Region)
            .ToDictionary(g => g.Key,
                g=>g.SelectMany(c => c.Orders).Where(
                    o => o.Total == g.SelectMany(c => c.Orders).Max(o => o.Total)).ToList());
        return customerOrdersTotalMax;
    }
}

File: Parallel-Java/src/parallel/linq101/mylinq/MaxMethods.java
package parallel.linq101.mylinq;

import org.javatuples.Pair;
import parallel.linq101.datasource.Customer;
import parallel.linq101.datasource.Data;
import parallel.linq101.datasource.Order;
import parallel.linq101.datasource.Product;

import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static java.util.Comparator.*;
import static java.util.stream.Collectors.*;

public class MaxMethods {
    //original:4 target:40
    public List<Product> getProductList() {
        return Data.getProductList();
    }

    public List<Customer> getCustomerList() {
        return Data.getCustomerList();
    }
    //max syntax : 5

    public int numbersMax() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int maxNum = Arrays.stream(numbers).max().getAsInt();
        return maxNum;
    }

    public int bigNumbersMax() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int maxNum = Arrays.stream(numbers).filter(n -> n >= 5).max().getAsInt();
        return maxNum;
    }

    public int oddNumbersMax() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int maxNum = Arrays.stream(numbers).filter(n -> n % 2 == 0).max().getAsInt();
        return maxNum;
    }

    public double doubleMax() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        double maxNum = Arrays.stream(numbers).max().getAsDouble();
        return maxNum;
    }

    public double bigDoubleNumbersMax() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        double maxNum = Arrays.stream(numbers).filter(n -> n >= 5).max().getAsDouble();
        return maxNum;
    }


    //max projection: 15

    public int numbersMaxTwice() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int maxNum = Arrays.stream(numbers).map(n -> n * 2).max().getAsInt();
        return maxNum;
    }

    public int bigNumbersMaxTwice() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int maxNum = Arrays.stream(numbers).filter(n -> n >= 5).map(n -> n * 2).max().getAsInt();
        return maxNum;
    }

    public int oddNumbersMaxTwice() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int maxNum = Arrays.stream(numbers).filter(n -> n % 2 == 0).map(n -> n * 2).max().getAsInt();
        return maxNum;
    }

    public int stringMaxLength() {
        String[] words = {"cherry", "apple", "blueberry"};
        int maxLength = Arrays.stream(words).mapToInt(w -> w.length()).max().getAsInt();
        return maxLength;
    }

    public int shortStringMaxLength() {
        String[] words = {"cherry", "apple", "blueberry"};
        int maxLength = Arrays.stream(words).filter(w -> w.length() < 7).mapToInt(w -> w.length()).max().getAsInt();
        return maxLength;
    }

    public double productUnitPriceMax() {
        List<Product> products = getProductList();
        double maxUnitPrice = products.stream().mapToDouble(p -> p.getUnitPrice()).max().getAsDouble();
        return maxUnitPrice;
    }

    public double expensiveProductUnitPriceMax() {
        List<Product> products = getProductList();
        double maxUnitPrice = products.stream().filter(p -> p.getUnitPrice() > 100)
                .mapToDouble(p -> p.getUnitPrice()).max().getAsDouble();
        return maxUnitPrice;
    }

    public int expensiveProductStockMax() {
        List<Product> products = getProductList();
        int maxUnitPrice = products.stream().filter(p -> p.getUnitPrice() > 100)
                .mapToInt(p -> p.getUnitsInStock()).max().getAsInt();
        return maxUnitPrice;
    }

    public int productLongNameLengthMax() {
        List<Product> products = getProductList();
        int maxUnitPrice = products.stream().map(p -> p.getProductName())
                .filter(name -> name.length() > 7).mapToInt(name -> name.length()).max().getAsInt();
        return maxUnitPrice;
    }

    public List<Pair<String, Double>> customerOrdersTotalMax() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersTotalMax = customers.stream().map(c -> Pair.with(c.getCustomerId(),
                c.getOrders().stream().mapToDouble(o -> o.getTotal()).max().getAsDouble())).toList();
        return ordersTotalMax;
    }

    public Map<String, Double> customerOrdersTotalMaxDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersTotalMax = customers.stream().collect(toMap(c -> c.getCustomerId(),
                c -> c.getOrders().stream().mapToDouble(o -> o.getTotal()).max().getAsDouble()));
        return ordersTotalMax;
    }

    public List<Pair<String, Double>> customerInRegionOrdersTotalMax(String region) {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersTotalMax = customers.stream().filter(c -> c.getRegion().equals(region))
                .map(c -> Pair.with(c.getCustomerId(),
                        c.getOrders().stream().mapToDouble(o -> o.getTotal()).max().getAsDouble())).toList();
        return ordersTotalMax;
    }

    public Map<String, Double> customerInRegionOrdersTotalMaxDict(String region) {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersTotalMax = customers.stream().filter(c -> c.getRegion().equals(region))
                .collect(toMap(c -> c.getCustomerId(),
                        c -> c.getOrders().stream().mapToDouble(o -> o.getTotal()).max().getAsDouble()));
        return ordersTotalMax;
    }

    public List<Pair<String, Double>> customerExpensiveOrdersTotalMax() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersTotalMax = customers.stream().map(c -> Pair.with(c.getCustomerId(),
                c.getOrders().stream().filter(o -> o.getTotal() > 100)
                        .mapToDouble(o -> o.getTotal()).max().getAsDouble())).toList();
        return ordersTotalMax;
    }

    public Map<String, Double> customerExpensiveOrdersTotalMaxDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersTotalMax = customers.stream().collect(toMap(c -> c.getCustomerId(),
                c -> c.getOrders().stream().filter(o -> o.getTotal() > 100)
                        .mapToDouble(o -> o.getTotal()).max().getAsDouble()));
        return ordersTotalMax;
    }

    //max grouped: 10

    public List<Pair<String, Integer>> productEachCategoryStock() {
        List<Product> products = getProductList();
        List<Pair<String, Integer>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().mapToInt(p -> p.getUnitsInStock()).max().getAsInt())).toList();
        return categories;
    }

    public Map<String, Integer> productEachCategoryStockDict() {
        List<Product> products = getProductList();
        Map<String, Integer> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().mapToInt(p -> p.getUnitsInStock()).max().getAsInt()));
        return categories;
    }

    public List<Pair<String, Integer>> expensiveProductEachCategoryStock() {
        List<Product> products = getProductList();
        List<Pair<String, Integer>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().filter(
                        p -> p.getUnitPrice() > 100).mapToInt(p -> p.getUnitsInStock()).max().getAsInt())).toList();
        return categories;
    }

    public Map<String, Integer> expensiveProductEachCategoryStockDict() {
        List<Product> products = getProductList();
        Map<String, Integer> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().filter(
                        p -> p.getUnitPrice() > 100).mapToInt(p -> p.getUnitsInStock()).max().getAsInt()));
        return categories;
    }

    public List<Pair<String, Integer>> customerEachRegionOrdersMax() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Integer>> ordersMax = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().mapToInt(c -> c.getOrders().size()).max().getAsInt())).toList();
        return ordersMax;
    }

    public Map<String, Integer> customerEachRegionOrdersMaxDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Integer> ordersMax = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().mapToInt(c -> c.getOrders().size()).max().getAsInt()));
        return ordersMax;
    }

    public List<Pair<String, Integer>> customerEachCountryOrdersMax() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Integer>> ordersMax = customers.stream().collect(groupingBy(c -> c.getCountry()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().mapToInt(c -> c.getOrders().size()).max().getAsInt())).toList();
        return ordersMax;
    }

    public Map<String, Integer> customerEachCountryOrdersMaxDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Integer> ordersMax = customers.stream().collect(groupingBy(c -> c.getCountry()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().mapToInt(c -> c.getOrders().size()).max().getAsInt()));
        return ordersMax;
    }

    public List<Pair<String, Double>> customerEachRegionOrdersTotalMax() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersTotalMax = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().flatMap(c ->
                        c.getOrders().stream()).mapToDouble(o -> o.getTotal()).max().getAsDouble())).toList();
        return ordersTotalMax;
    }

    public Map<String, Double> customerEachRegionOrdersTotalMaxDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersTotalMax = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().flatMap(c ->
                        c.getOrders().stream()).mapToDouble(o -> o.getTotal()).max().getAsDouble()));
        return ordersTotalMax;
    }

    //max each group: 10
    public List<Pair<String, List<Product>>> productEachCategoryMaxUnitPriceProducts() {
        List<Product> products = getProductList();
        List<Pair<String, List<Product>>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().collect(groupingBy(p -> p.getUnitPrice()))
                        .entrySet().stream().max(Map.Entry.comparingByKey()).get().getValue())).toList();
        return categories;
    }

    public Map<String, List<Product>> productEachCategoryMaxUnitPriceProductsDict() {
        List<Product> products = getProductList();
        Map<String, List<Product>> categories = products.stream().collect(groupingBy(p -> p.getCategory())).entrySet()
                .stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().collect(groupingBy(
                        p -> p.getUnitPrice())).entrySet().stream().max(Map.Entry.comparingByKey()).get().getValue()));
        return categories;
    }

    public List<Pair<String, List<Product>>> expensiveProductEachCategoryMaxUnitPriceProducts() {
        List<Product> products = getProductList();
        List<Pair<String, List<Product>>> categories = products.stream().filter(p -> p.getUnitPrice() > 100).
                collect(groupingBy(p -> p.getCategory())).entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().collect(groupingBy(p -> p.getUnitPrice()))
                                .entrySet().stream().max(Map.Entry.comparingByKey()).get().getValue())).toList();
        return categories;
    }

    public Map<String, List<Product>> expensiveProductEachCategoryMaxUnitPriceProductsDict() {
        List<Product> products = getProductList();
        Map<String, List<Product>> categories = products.stream().filter(p -> p.getUnitPrice() > 100).collect(
                groupingBy(p -> p.getCategory())).entrySet().stream().collect(toMap(g -> g.getKey(),
                g -> g.getValue().stream().collect(groupingBy(p -> p.getUnitPrice()))
                        .entrySet().stream().max(Map.Entry.comparingByKey()).get().getValue()));
        return categories;
    }

    public List<Pair<String, List<Customer>>> customerEachRegionOrdersMaxCustomer() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, List<Customer>>> customerOrdersMax = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().
                        collect(groupingBy(c -> c.getOrders().size())).entrySet()
                        .stream().max(Map.Entry.comparingByKey()).get().getValue())).toList();
        return customerOrdersMax;
    }

    public Map<String, List<Customer>> customerEachRegionOrdersMaxCustomerDict() {
        List<Customer> customers = getCustomerList();
        Map<String, List<Customer>> customerOrdersMax = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().
                        collect(groupingBy(c -> c.getOrders().size())).entrySet()
                        .stream().max(Map.Entry.comparingByKey()).get().getValue()));
        return customerOrdersMax;
    }

    public List<Pair<String, List<Customer>>> customerEachCountryOrdersMaxCustomer() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, List<Customer>>> customerOrdersMax = customers.stream().collect(groupingBy(c -> c.getCountry()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().
                        collect(groupingBy(c -> c.getOrders().size())).entrySet()
                        .stream().max(Map.Entry.comparingByKey()).get().getValue())).toList();
        return customerOrdersMax;
    }

    public Map<String, List<Customer>> customerEachCountryOrdersMaxCustomerDict() {
        List<Customer> customers = getCustomerList();
        Map<String, List<Customer>> customerOrdersMax = customers.stream().collect(groupingBy(c -> c.getCountry()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().
                        collect(groupingBy(c -> c.getOrders().size())).entrySet()
                        .stream().max(Map.Entry.comparingByKey()).get().getValue()));
        return customerOrdersMax;
    }

    public List<Pair<String, List<Order>>> customerEachRegionOrdersTotalMaxCustomer() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, List<Order>>> customerOrdersTotalMax = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream()
                        .flatMap(c -> c.getOrders().stream()).collect(groupingBy(o -> o.getTotal()))
                        .entrySet().stream().max(Map.Entry.comparingByKey()).get().getValue())).toList();
        return customerOrdersTotalMax;
    }

    public Map<String, List<Order>> customerEachRegionOrdersTotalMaxCustomerDict() {
        List<Customer> customers = getCustomerList();
        Map<String, List<Order>> customerOrdersTotalMax = customers.stream().collect(groupingBy(c -> c.getCountry()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream()
                        .flatMap(c -> c.getOrders().stream()).collect(groupingBy(o -> o.getTotal()))
                        .entrySet().stream().max(Map.Entry.comparingByKey()).get().getValue()));
        return customerOrdersTotalMax;
    }
}
